### DISCUSSION TITLE
XIAO ESP32C3 accesses Home Assistant via ESPHome service | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#108] [Q&A] [answered]
# XIAO ESP32C3 accesses Home Assistant via ESPHome service | Seeed Studio Wiki

Accessing ESPHome with XIAO ESP32C3

https://wiki.seeedstudio.com/xiao-esp32c3-esphome/

**Comment by djwdjw321** [comment]
I can't get this to work. After a lot of effort I have found that the xiao is sending some "garbage" to the radar sensor on boot up. It also seems that the sensor hangs and doesn't send data if it receives data before it is fully booted up. There may be a fix in a later version of the sensor firmware but I can't upgrade the firmware on my sensors, as the version I have doesn't support serial update, only updating using an expensive programmer that I don't have. I haven't yet found a way to stop the xiao sending the unwanted bytes and am still working on it. I can get it to work by powering up the sensor first and then, after a short delay, powering up the xiao. This is not a practical solution as the sensor is being powered from the same supply as the xiao. Anyone have any ideas on how to suppress the unwanted bytes from the xiao?

**Comment by Tsamsiyu1** [comment]
I finnally got my sensors. unfortunatly when i do it like this my ESPHome device has no entities. Does anybody has a idea why?

**Reply by Tsamsiyu1** [reply]
installed it a couple more times and i see the entities but still nothing works. i cant change any values then all stay at 0.

**Comment by Tsamsiyu1** [comment]
this manual has an error. it says connect D7 to TX and D6 to RX but in the  xiaoesp32c3-mr24hpc1_part2.yaml it says  20 (D7) is RX and 21 (D6) is TX.
But even with sodering it the other way arround i get :
[15:26:42][D][main:334]: text_sensor on_raw_value
[15:26:42][D][text_sensor:067]: 'Standard Product model': Sending state ''
[15:27:12][D][text_sensor:067]: 'Standard Product ID': Sending state ''
[15:27:28][D][text_sensor:067]: 'Standard protocol type': Sending state ''
[15:27:34][D][text_sensor:067]: 'Standard Firmware version': Sending state ''
[15:27:35][D][text_sensor:067]: 'Standard Hardware model': Sending state ''

**Reply by rcruikshank** [reply]
I think RX goes to TX and visa versa. Crossover is normal for RX/TX.

**Comment by rcruikshank** [comment]
Only way I get it to work is to power the mmWave board up after the ESP32C3 MCU.  I put a switch between the MCU and the 5v terminal on the mmWave board and then power the MCU and then switch on the mmWave board and it works.

**Comment by limengdu** [comment]
https://wiki.seeedstudio.com/xiao-esp32c3-esphome/#faq4-my-xiao-esp32c3-is-connected-to-linkstars-network-but-why-dont-i-see-the-sensor-data-refreshed

**Comment by rcruikshank** [comment, marked as answer]
Perhaps you can include a delay circuit so the mmWave board doesn't power up at the same time as the ESP32C3 MCU.  Put a C capacitor between MCU GND and the mmWave board and a R resistor between the MCU 5v and the mmWave board. 

τ = R × C
5 seconds = R × 1000μF
R = 5 seconds / 1000μF
R = 5000 ohms

         +5V
          |
          |
          R
          |
          |
          +---> Load
          |
          |
          C
          |
          |
         GND

In this circuit diagram, the 5V power supply is connected to a resistor (R) and a capacitor (C) in series. The load is connected to the junction of the resistor and capacitor. When power is applied, the capacitor begins to charge through the resistor. The time it takes for the capacitor to charge up to the supply voltage depends on the values of the resistor and capacitor. In this circuit, the values of the resistor and capacitor have been chosen to provide a delay of approximately 5 seconds. After the capacitor has charged up, it provides power to the load, which can be a sensor or other electronic device. Note that the actual delay time may vary slightly due to the tolerances of the components used.

So a 5k ohm resistor and 1000μF capacitor would delay the mmWave powering up for 5 second.  Might be enough?  Would this work?

**Reply by rcruikshank** [reply]
Nope doesn't seem to work. mmWave needs to be powered up then down then up again before it will work?

**Reply by Tsamsiyu1** [reply]
i still get the same  emptz response from the sensor

**Reply by rcruikshank** [reply]
even when you pull the 5v from the mmWave board and reconnect it?

**Reply by rcruikshank** [reply]
After powering the mmWave board on again you then have to turn the board on in the Home Assistant system via the "Custom Info Output" switch.

**Reply by djwdjw321** [reply]
OK I have found a fix for this. I can't do a sensor firmware update as I don't have a J-Link device and a firmware upgrade may not even do the trick. I worked on suppressing the bootup messages being sent out on the serial link to the sensor.  This may be a possible solution and I was able to suppress some messages but still had a long way to go. Finally, I tried a hardware solution and it works. The technique is to disable the serial transmit from the Xiao to the 24ghz sensor for a few seconds on boot up. I did this by putting a 2.2kohm resistor in series with the serial connection (xiao TX to sensor RX) to allow the sensor input to be pulled high (high is off for serial) without damaging the Xiao. Then I used a GPIO pin (I used GPIO3 - D1 on the Xiao-C3) with a 47k pulldown to 0v (internal pulldowns are not enabled on boot up). Then a PNP transistor was connected as follows: Emitter to 3.3v (NOT 5v) base to the GPIO pin and collector to the sensor RX input. Then when GPIO3 is low/off, the base of the transistor is pulled low and the serial pin is pulled high and disabled. When the GPIO pin is is high/on the transistor is turned off and the serial link is enabled.

I then added a gpio switch and a "on_bootup" section to the YAML to start with the GPIO off and after 3 seconds turn it on. This fixed the problem. The YAML looks as follows:

In the switch: section add:-

 - platform: gpio
    name: "serial_enable"
    id: serial_enable
    internal: true
    restore_mode: ALWAYS_OFF
    pin:
      number: 3
      inverted: false
      mode:
        output: true
        pullup: false

In the esphome: section add:-

 on_boot:
    priority: 800
    then:
      - switch.turn_off: serial_enable
      - delay: 3s
      - switch.turn_on: serial_enable

simpler YAML may work but this did the trick.

**Reply by MatthewJeffson** [reply]
wow, amazing work! :o

**Reply by rcruikshank** [reply]
Fantastic.  One question:
When you say "GPIO3 - D1 on the Xiao-C3 with a 47k pulldown to 0v (internal pulldowns are not enabled on boot up)" what do you mean 0v? 47k resistor between GPIO3 and GND?

**Reply by djwdjw321** [reply]
Yes. The 47k goes between the GPIO (GPIO3) and GND. Attached is a picture of my sensor with the components installed. The radar sensor is below the xiao. Also note that the YAML configuration supplied in the wiki is incorrect in a number of ways, such as "custom mode settings" is defined as a number with 250 values instead of 4 values (1,2,3,4). It works but I am going over all the YAML to clean it up.


![20230404_134231](https://user-images.githubusercontent.com/5524540/229923111-db99e767-96fc-48da-a44e-97e86419e19e.jpg)

**Reply by rcruikshank** [reply]
Something like this:
![circuit](https://user-images.githubusercontent.com/47944823/230241068-8e69069e-5806-499f-92ff-2736651ff396.png)

I get a couple of errors:
![errors](https://user-images.githubusercontent.com/47944823/230241140-c8c41895-4c88-4fd6-b04c-49544925c2b4.png)

Also I note that you can flash the sensor via [UART](https://wiki.seeedstudio.com/Radar_MR24HPC1/#update-firmware-via-uart). I wonder if the firmware needs upgrading. I think you can tell what version of firmware is installed also via this method.  This would be simple to do. Might give it a go.

**Reply by rcruikshank** [reply]
ok so I connected the mmWave sensor to a [5v UART to USB flasher](https://www.ebay.com.au/itm/392269284896?hash=item5b55124620:g:~ycAAOSwpUVdROvg&amdata=enc%3AAQAHAAAA4Pz12spczdbhYI9Hc7L4pmQ0F3YFg%2BeWg%2BwOhxYB9F%2FoZOuED2IzWVY8e665lab6XTX8lYZF9hjVKiowW5MkSghtcNZiqMyK5QCoa2JEQ9BPMEBy9dy0CpMHt60M%2BkRiKom%2Bq91u3MN4dm3aQOH8lLlxQQaMhaVuEpbfLP6eQZnkEHec2r%2FOEaNwUUzV3cXlaJ79WwvV3%2BN3gRrC0xnI2tSVHQyDNmehU6TCFTZRGHKFIkhoHYcsxet3icEQlxpujSx7GW15HlD%2Fh0lt54buzggYk7mOz4opHtl9v8xxj6No%7Ctkp%3ABFBM1ujPzeph) and checked the installed firmware version as described [here](https://wiki.seeedstudio.com/Radar_MR24HPC1/#update-firmware-via-uart).  

Result: 
![mmWavecapture](https://user-images.githubusercontent.com/47944823/230273371-d7aae1cb-d19f-4516-adb3-2aca6ff53219.png)

Using a Hex editor:
![mmWaveVersion](https://user-images.githubusercontent.com/47944823/230273630-b63c96b2-bca1-48f4-baaa-f15f8a9b09f6.png)

Looks like version G24VD1SYV001006.
 
They warn the the following:
_"Make sure your radar firmware version is at least version G24VD1SYV001006 before using the UART to upgrade the firmware, otherwise it may disable the radar, at which point you'll have to use J-link to burn the firmware to use it!"_

So I upgraded it with the bin file UART_MR24HPC1-20230302.bin but the version number didn't seem to change.  Its still G24VD1SYV001006?  Also no change in performance.  Still have to disconnect 5v from sensor after powering up ESP32C3 to get it to work!

**Reply by rcruikshank** [reply]
> OK I have found a fix for this. I can't do a sensor firmware update as I don't have a J-Link device and a firmware upgrade may not even do the trick. I worked on suppressing the bootup messages being sent out on the serial link to the sensor. This may be a possible solution and I was able to suppress some messages but still had a long way to go. Finally, I tried a hardware solution and it works. The technique is to disable the serial transmit from the Xiao to the 24ghz sensor for a few seconds on boot up. I did this by putting a 2.2kohm resistor in series with the serial connection (xiao TX to sensor RX) to allow the sensor input to be pulled high (high is off for serial) without damaging the Xiao. Then I used a GPIO pin (I used GPIO3 - D1 on the Xiao-C3) with a 47k pulldown to 0v (internal pulldowns are not enabled on boot up). Then a PNP transistor was connected as follows: Emitter to 3.3v (NOT 5v) base to the GPIO pin and collector to the sensor RX input. Then when GPIO3 is low/off, the base of the transistor is pulled low and the serial pin is pulled high and disabled. When the GPIO pin is is high/on the transistor is turned off and the serial link is enabled.
> 
> I then added a gpio switch and a "on_bootup" section to the YAML to start with the GPIO off and after 3 seconds turn it on. This fixed the problem. The YAML looks as follows:
> 
> In the switch: section add:-
> 
> - platform: gpio
>   name: "serial_enable"
>   id: serial_enable
>   internal: true
>   restore_mode: ALWAYS_OFF
>   pin:
>     number: 3
>     inverted: false
>     mode:
>       output: true
>       pullup: false
> 
> In the esphome: section add:-
> 
> on_boot: priority: 800 
>   then: 
>     - switch.turn_off: serial_enable 
>     - delay: 3s 
>     - switch.turn_on: serial_enable
> 
> simpler YAML may work but this did the trick.

I get the following error when compiling this yaml:
```
Linking /data/diningroom-presence/.pioenvs/diningroom-presence/firmware.elf
/data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /data/diningroom-presence/.pioenvs/diningroom-presence/src/main.o: in function `setup()':
/config/esphome/diningroom-presence.yaml:398: undefined reference to `vtable for esphome::gpio::GPIOSwitch'
/data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /config/esphome/diningroom-presence.yaml:398: undefined reference to `vtable for esphome::gpio::GPIOSwitch'
/data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /config/esphome/diningroom-presence.yaml:398: undefined reference to `vtable for esphome::gpio::GPIOSwitch'
/data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /config/esphome/diningroom-presence.yaml:398: undefined reference to `vtable for esphome::gpio::GPIOSwitch'
collect2: error: ld returned 1 exit status
*** [/data/diningroom-presence/.pioenvs/diningroom-presence/firmware.elf] Error 1
```

**Comment by sawo69** [comment]
I get these errors while compiling the software. What can I do about them?

In file included from src/main.cpp:112:
src/R24dvd.h: In member function 'void UartReadLineSensor::R24_frame_parse_human_information(uint8_t*)':
src/R24dvd.h:726:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 3 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h:742:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 9 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h:749:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 3 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h:762:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 3 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h:777:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 9 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h:784:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 3 && data[FRAME_DATA_INDEX] >= 0) {
                                           ~~~~~~~~~~~~~~~~~~~~~~~^~~~
src/R24dvd.h: In member function 'void UartReadLineSensor::R24_frame_parse_open_underlying_information(uint8_t*)':
src/R24dvd.h:918:66: warning: comparison is always true due to limited range of data type [-Wtype-limits]
         if (data[FRAME_DATA_INDEX] < 3 && data[FRAME_DATA_INDEX] >= 0) {

**Reply by rcruikshank** [reply]
Did you copy the [R24dvd.h](https://files.seeedstudio.com/wiki/homs-xiaoc3-linkstar/R24dvd.h) file into the `\\[yourhomeassistant].local\config\esphome\` directory?

**Reply by sawo69** [reply]
Yes i double checked and the file is there.

**Reply by rcruikshank** [reply]
What's in your yaml code in esphome?
The warnings is due to the comparison of an unsigned integer with a signed integer. The comparison is always true because an unsigned integer cannot be negative, so the second part of the comparison is redundant. You can fix this by changing the type of the variable being compared to an unsigned integer or by removing the second part of the comparison.

**Reply by djwdjw321** [reply]
These warning in R24dvd.h are unimportant but I removed the second (unnecessary - because an unsigned int can never be negative) part of the comparisons just to clean up the code. The YAML does work but the starting sequence issue made me think it didn't for a long time. Now I have fixed the starting sequence issue with the hardware mod, I am starting to look at the YAML and it looks like something thrown together just to test the system. The radar sensor user guide defines the various messages and responses and I am in the process of modifying the YAML to match the user guide.

**Comment by MatthewJeffson** [comment]
Welcome to our product documentation platform! We value user feedback and contributions, so we couldn’t be more grateful that you can help us improving the documentation.

Hence, if you find obvious errors in the product documentation, or you have additional ideas and insights about using products, please [contribute](https://wiki.seeedstudio.com/Contribution-Guide) to us.

If you need technical support, you can email it to [fae@seeed.cc](mailto:fae@seeed.cc).

If you want to discuss the product, I recommend the [Seeed Studio Forum](https://forum.seeedstudio.com/) and [Seeed Studio Discord Channel](https://discord.gg/eWkprNDMU7)

The "Discussion channel" in Github here is a relatively new addition and may not receive much attention.

Nonetheless, we hope you find it helpful with the help of others. We will clean it up on a monthly basis.

**Comment by gabberhq** [comment]
got some problem when trying to flash :

INFO ESPHome 2023.12.3
INFO Reading configuration /config/esphome/xiao-esp32c3.yaml...
INFO Updating https://github.com/limengdu/mmwave-kit-external-components.git@main
Failed config

binary_sensor.seeed_mr24hpc1: [source <unicode string>:90]
  platform: seeed_mr24hpc1
  
  [someone_exist] is an invalid option for [binary_sensor.seeed_mr24hpc1]. Please check the indentation.
  someone_exist: 
    name: Presence Information
button.seeed_mr24hpc1: [source <unicode string>:117]
  platform: seeed_mr24hpc1
  
  [Reset] is an invalid option for [button.seeed_mr24hpc1]. Please check the indentation.
  Reset: 
    name: Module Reset
  custom_set_end: 
    name: End Of Custom Mode Settings

**Comment by gabberhq** [comment]
Please correct the source code :

reset:     should be =>   restart:

someone_exist:     should be =>  has_target:

**Comment by Stebar59** [comment]
I have following errors when tryin to istall the configuration to the XIAO ESP32C3 using esphome:

Successfully installed cachecontrol-0.13.1 certifi-2023.11.17 cffi-1.16.0 charset-normalizer-3.3.2 click-8.1.7 colorama-0.4.6 contextlib2-21.6.0 cryptography-3.4.8 filelock-3.13.1 future-0.18.3 idf-component-manager-1.4.1 idna-3.6 kconfiglib-13.7.1 msgpack-1.0.7 packaging-23.2 pycparser-2.21 pyparsing-2.3.1 pyyaml-6.0.1 requests-2.31.0 requests-file-1.5.1 requests-toolbelt-1.0.0 schema-0.7.5 six-1.16.0 tqdm-4.66.1 urllib3-1.26.18
Reading CMake configuration...
-- Found Git: /usr/bin/git (found version "2.39.2") 
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc -- broken
-- Configuring incomplete, errors occurred!
See also "/data/build/xiao-esp32c3/.pioenvs/xiao-esp32c3/CMakeFiles/CMakeOutput.log".
See also "/data/build/xiao-esp32c3/.pioenvs/xiao-esp32c3/CMakeFiles/CMakeError.log".

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/tool-cmake/share/cmake-3.16/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/build/xiao-esp32c3/.pioenvs/xiao-esp32c3/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/data/cache/platformio/packages/tool-ninja/ninja cmTC_4ea93 && [1/2] Building C object CMakeFiles/cmTC_4ea93.dir/testCCompiler.c.obj
    [2/2] Linking C executable cmTC_4ea93
    FAILED: cmTC_4ea93 
    : && /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc -march=rv32imc  -nostartfiles -march=rv32imc --specs=nosys.specs CMakeFiles/cmTC_4ea93.dir/testCCompiler.c.obj  -o cmTC_4ea93   && :
    /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../lib/gcc/riscv32-esp-elf/8.4.0/../../../../riscv32-esp-elf/bin/ld: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../libexec/gcc/riscv32-esp-elf/8.4.0/liblto_plugin.so: error loading plugin: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/../libexec/gcc/riscv32-esp-elf/8.4.0/liblto_plugin.so: cannot open shared object file: No such file or directory
    collect2: error: ld returned 1 exit status
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf/tools/cmake/project.cmake:298 (__project)
  CMakeLists.txt:3 (project)



========================= [FAILED] Took 608.66 seconds =========================


Someone has an idea?
Thanks

Stefano
